substrate/client/network/src/litep2p/peerstore.rs

```diff
@@ -0,0 +1,396 @@
+/// `i32::MAX` becomes 0 in ~1100 seconds (actually less due to integer arithmetic).
+const INVERSE_DECREMENT: i32 = 50;
+
+/// Amount of time between the moment we last updated the [`PeerStore`] entry and the moment we
+/// remove it, once the reputation value reaches 0.
+const FORGET_AFTER: Duration = Duration::from_secs(3600);
+
+/// Peer information.
+#[derive(Debug, Clone, Copy)]
+struct PeerInfo {
+	/// Reputation of the peer.
+	reputation: i32,
+
+	/// Instant when the peer was last updated.
+	last_updated: Instant,
+
+	/// Role of the peer, if known.
+	role: Option<ObservedRole>,
+}
+
+impl Default for PeerInfo {
+	fn default() -> Self {
+		Self { reputation: 0i32, last_updated: Instant::now(), role: None }
+	}
+}
+
+impl PeerInfo {
+	fn is_banned(&self) -> bool {
+		self.reputation < BANNED_THRESHOLD
+	}
+
+	fn decay_reputation(&mut self, seconds_passed: u64) {
+		// Note that decaying the reputation value happens "on its own",
+		// so we don't do `bump_last_updated()`.
+		for _ in 0..seconds_passed {
+			let mut diff = self.reputation / INVERSE_DECREMENT;
+			if diff == 0 && self.reputation < 0 {
+				diff = -1;
+			} else if diff == 0 && self.reputation > 0 {
+				diff = 1;
+			}
+
+			self.reputation = self.reputation.saturating_sub(diff);
+
+			if self.reputation == 0 {
+				break
+			}
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+pub struct PeerstoreHandleInner {
+	peers: HashMap<PeerId, PeerInfo>,
+	protocols: Vec<Arc<dyn ProtocolHandle>>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct PeerstoreHandle(Arc<Mutex<PeerstoreHandleInner>>);
+
+impl PeerstoreHandle {
+	/// Add known peer to [`Peerstore`].
+	pub fn add_known_peer(&self, peer: PeerId) {
+		self.0
+			.lock()
+			.peers
+			.insert(peer, PeerInfo { reputation: 0i32, last_updated: Instant::now(), role: None });
+	}
+
+	pub fn peer_count(&self) -> usize {
+		self.0.lock().peers.len()
+	}
+
+	fn progress_time(&self, seconds_passed: u64) {
+		if seconds_passed == 0 {
+			return
+		}
+
+		let mut lock = self.0.lock();
+
+		// Drive reputation values towards 0.
+		lock.peers
+			.iter_mut()
+			.for_each(|(_, info)| info.decay_reputation(seconds_passed));
+
+		// Retain only entries with non-zero reputation values or not expired ones.
+		let now = Instant::now();
+		lock.peers
+			.retain(|_, info| info.reputation != 0 || info.last_updated + FORGET_AFTER > now);
+	}
+}
+
+impl PeerStoreProvider for PeerstoreHandle {
+	fn is_banned(&self, peer: &PeerId) -> bool {
+		self.0.lock().peers.get(peer).map_or(false, |info| info.is_banned())
+	}
+
+	/// Register a protocol handle to disconnect peers whose reputation drops below the threshold.
+	fn register_protocol(&self, protocol_handle: Arc<dyn ProtocolHandle>) {
+		self.0.lock().protocols.push(protocol_handle);
+	}
+
+	/// Report peer disconnection for reputation adjustment.
+	fn report_disconnect(&self, _peer: PeerId) {
+		unimplemented!();
```

Comment:
The function doesn't serve much purpose beyond adjusting reputation so `Peerset` just directly calls `Peerstore::report_peer()` when a peer disconnects. I can start using this function if you think it's preferable but because this function is not used, disconnections and open failures can share the same code.

Reply:
May be remove it completely then?

Reply:
It's part of the `PeerStoreProvider` trait which is used by the other `PeerStore`. I can refactor the backends to use the same `PeerStore` in a follow-up and figure out what to do with this function